### DISCUSSION TITLE
Gauge: accessing valBits

### DISCUSSION THREAD

**Discussion by skwiffhub** [prometheus/client_golang#958] [Q&A] [answered]
Noob question: Is there any particular reason that `valBits` would not be exposed to a caller via `Gauge`? Seems like pretty basic counter functionality to simply return its current value.

`valBits` source ref: https://github.com/prometheus/client_golang/blob/master/prometheus/gauge.go#L82

**Comment by beorn7** [comment]
Generally, the various metric types are specialized on updating their value in instrumentation and then return them in a scrape. If we had methods to directly access the metric values, it would signal that those methods are useful and efficient, and generally meant to be used for regular operation.

See also https://pkg.go.dev/github.com/prometheus/client_golang@v1.11.0/prometheus/testutil#ToFloat64 for a helper function to access the value of a Gauge and why that's only recommended for special cases like tests.

**Reply by skwiffhub** [reply]
I see your point. However, there is the general issue of synchronization for which a quick check of `valBits` would provide a sanity check. For example, a `Gauge` that goes negative when the user intends for it to always be a positive integer. Alternatively, a `Gauge` that exceeds a pre-defined maximum. These are all most likely software errors, but nonetheless it would be useful to be able to access the value and re-calibrate it or log an error.

Perhaps an API with an accompanying set of disclaimers?

**Comment by beorn7** [comment, marked as answer]
In that case, I would recommend to manage the Gauge value on your own and expose it via https://pkg.go.dev/github.com/prometheus/client_golang@v1.11.0/prometheus#NewGaugeFunc

Note that you run into all kind of concurrency problems. After retrieving the value from a Gauge, another goroutine might already update it before you set it again. Furthermore, future versions of client_golang might leverage core-local variables that are only gathered together upon scrape…
